Fix #326 Migrate to using Maven 3 APIs and Aether for dependency resolution

srdo opened the pull request:
Based on https://github.com/mojohaus/license-maven-plugin/issues/324, please look at that one first.

This migrates the plugin to Maven 3 dependencies, and also replaces the old ArtifactResolver with maven-resolver.

The version requirement here is at least 3.5.0, as maven-resolver-provider is not present in prior versions. For now I've set the plugin to need 3.6.1, as that was what I tested with.

srdo commented:
Set required Maven version to 3.5.4, as that is the version used by CI.

ppalaga commented:
> Set required Maven version to 3.5.4, as that is the version used by CI.

+1

ppalaga commented on src/main/java/org/codehaus/mojo/license/api/DefaultThirdPartyTool.java:
```diff
@@ -754,59 +750,80 @@ private void loadOneGlobalSet( SortedSet<MavenProject> unsafeDependencies, Licen
+            String type, String classifier, List<RemoteRepository> remoteRepositories )
+                    throws ArtifactResolutionException
     {
         // TODO: this is a bit crude - proper type, or proper handling as metadata rather than an artifact in 2.1?
```
Shouldn't this comment be removed? It does not seem to make any sense anymore.

srdo replied:
Maybe? I understood the comment to say that handling the third party descriptor as an artifact was crude, but I'm not sure what the intended alternative was. I'll remove the comment.

srdo replied:
Removed the comment

ppalaga commented on src/main/java/org/codehaus/mojo/license/api/DefaultThirdPartyTool.java:
```diff
@@ -837,4 +854,10 @@ private File resolveArtifact( String groupId, String artifactId, String version,
+    @Override
+    public void setAetherRepoSession( RepositorySystemSession aetherRepoSession )
+    {
+        this.aetherRepoSession = aetherRepoSession;
```
Making aetherRepoSession mutable in this way does not feel good. If there is no way to persuade plexus to inject it, I vote for removing the aetherRepoSession field and pass it via method parameter.

srdo replied:
I agree, it is not nice. I couldn't find a way to inject it. I'll move it to method parameters.

srdo replied:
Replaced with MavenSession injection

ppalaga commented on src/main/java/org/codehaus/mojo/license/api/DependenciesTool.java:
```diff
@@ -163,8 +171,25 @@
+                     */
+                    Properties props = new Properties();
+                    props.putAll( System.getProperties() );
+                    ProjectBuildingRequest req = new DefaultProjectBuildingRequest()
```
Do we need a new instance of ProjectBuildingRequest for each iteration of the enclosing for loop?

srdo replied:
The parameters don't change, so probably not. I'll change it to reuse the request.

srdo replied:
I think it is better here. Constructing the request shouldn't take any real time, and moving it outside the for loop puts a lot of distance between the declaration and use of the request.

ppalaga replied:
> Constructing the request shouldn't take any real time

It very much depends on how many iterations the loop has and also on how many sys props are set.

> puts a lot of distance

Sounds like a readability/aesthetics concern? - those are important, but here, I'd favor performance, unless proven that pulling the req out of the loop is technically impossible.

srdo replied:
Sure, but the system properties aren't being copied anymore, since the session change.

It's just a readability concern.  

I'll move it out of the loop.

srdo replied:
Done

ppalaga commented on src/main/java/org/codehaus/mojo/license/api/DependenciesTool.java:
```diff
@@ -163,8 +171,25 @@
+                     * Some POMs may refer to e.g. java.home or java.version,
+                     * so we need to ensure common system properties are set when parsing POMs.
+                     * Using the current system properties seems like a decent solution.
+                     */
```
I wonder if it wasn't safer to derive the ProjectBuildingRequest from the currently running request. Something like https://github.com/blackducksoftware/hub-maven-plugin/blob/master/src/main/java/com/blackducksoftware/integration/maven/MavenDependencyExtractor.java#L73

srdo replied:
That looks better, wasn't aware of MavenSession. Will try it out.

srdo replied:
Using MavenSession

ppalaga commented on src/main/java/org/codehaus/mojo/license/api/DependenciesTool.java:
```diff
@@ -239,4 +264,9 @@
+
+    public void setAetherRepoSession( RepositorySystemSession aetherRepoSession )
+    {
+        this.aetherRepoSession = aetherRepoSession;
```
Making aetherRepoSession mutable in this way does not feel good. If there is no way to persuade plexus to inject it, I vote for removing the aetherRepoSession field and pass it via method parameter.

srdo replied:
Replaced with MavenSession injection

ppalaga commented on src/main/java/org/codehaus/mojo/license/download/LicensedArtifactResolver.java:
```diff
@@ -216,4 +231,8 @@ public void loadProjectDependencies( ResolvedProjectDependencies artifacts,
 
+    public void setAetherRepoSession( RepositorySystemSession aetherRepoSession )
+    {
+        this.aetherRepoSession = aetherRepoSession;
```
Making aetherRepoSession mutable in this way does not feel good. If there is no way to persuade plexus to inject it, I vote for removing the aetherRepoSession field and pass it via method parameter.

srdo replied:
Replaced with MavenSession injection

srdo commented on src/main/java/org/codehaus/mojo/license/AbstractAddThirdPartyMojo.java:
```diff
@@ -872,6 +871,17 @@ public void setExcludedLicenses( String excludedLicenses ) throws MojoExecutionE
+     * Set repo session in the components that need it.
+     * @param session The repository system session
+     */
+    public void setAetherRepoSession( RepositorySystemSession session )
```
Maven calls it when injecting the RepositorySystemSession. The idea was to have this setter initialize the fields in the other classes. I couldn't find a way to get the DI framework to inject it.

srdo replied:
Replaced with MavenSession injection

srdo commented:
Addressed comments, and reworded commit message.

The MavenSession has a getter for the RepositorySystemSession, and can be injected with `@Requirement`, so I just did that rather than add RepositorySystemSession as a parameter everywhere.

srdo commented on src/main/java/org/codehaus/mojo/license/AbstractLicensesXmlMojo.java:
```diff
@@ -184,4 +183,14 @@ protected void writeLicenseSummary( List<ProjectLicenseInfo> deps, File licenses
+     * Set repo session in the components that need it.
+     * @param session The repository system session
+     */
+    public void setAetherRepoSession( RepositorySystemSession session )
```
Replaced with MavenSession injection

srdo commented on src/main/java/org/codehaus/mojo/license/AbstractThirdPartyReportMojo.java:
```diff
@@ -650,5 +644,15 @@ public String getEncoding()
+     * Set repo session in the components that need it.
+     * @param session The repository system session
+     */
+    public void setAetherRepoSession( RepositorySystemSession session )
```
Replaced with MavenSession injection

srdo commented:
The test failure is a timeout.

```
[WARNING] Unable to retrieve license from URL 'http://www.opensource.org/licenses/bsd-license.php' for dependency 'org.hamcrest:hamcrest-core:1.3': Read timed out
```

ppalaga commented:
Thanks, looks very good.
Sorry for the last nit pick: Could you please pull the buildReq out of the loop also in LicensedArtifactResolver.loadProjectDependencies()?

srdo commented:
Yes, done.

ppalaga commented:
Ouch, an editorconfig violation. Otherwise looks mergeable. Thanks!

ppalaga commented:
Great work, thanks a lot, @srdo !
